docs/tutorials/demo_best_practices.py

```diff
@@ -0,0 +1,271 @@
+We write a descriptive commit message.
+"""
+# %%
+with fs.transaction as tx:
+    df.to_csv(f"lakefs://{REPO_NAME}/main/weather_2010.csv")
+    commit = tx.commit(repository=REPO_NAME, branch="main", message="Preprocess 2010 data")
+print(commit)
+
+# %% [markdown]
+"""
+We see that the commit now has a unique id, the commit SHA which we can log to identify this particular state of the data.
+
+As the data looks good, we can also merge the branch back into `main`.
+"""
+
+# %%
+with fs.transaction as tx:
+    tx.merge(repository=REPO_NAME, source_ref=NEW_BRANCH_NAME, into="main")
+
+# %% [markdown]
+"""
+We will now start to develop our ML model. We recommend checking out a new branch for every set of experiments.
+Here, we will conduct the train test split and further experiment specific modifications, if applicable.
+"""
+
+# %%
+TRAINING_BRANCH = "experiment-1"
+with fs.transaction as tx:
+    tx.create_branch(repository=REPO_NAME, name=TRAINING_BRANCH, source="main")
+
+df = pd.read_csv(f"lakefs://{REPO_NAME}/{TRAINING_BRANCH}/weather_2010.csv")
+model_data = df.drop("time", axis=1)
+train, test = sklearn.model_selection.train_test_split(model_data, random_state=7)
+
+# %% [markdown]
+"""
+## Descriptive Tags for Human-Readability and Uniqe SHAs for unique Identification
+
+Since the train test split of the new branch is something we expect to address quite often in development, we will also add a human-readable tag to it.
+This makes it easy to refer back to it and to communicate this specific state of the data to colleagues. Tags are immutable which ensures consistency.
+You and your colleagues can then also work with (the same state (i.e. train test split, etc.) of the data.
+"""
+
+# %%
+TAG_NAME = "exp1_2010_data"
+with fs.transaction as tx:
+    train.to_csv(f"lakefs://{REPO_NAME}/{TRAINING_BRANCH}/train_weather.csv")
+    test.to_csv(f"lakefs://{REPO_NAME}/{TRAINING_BRANCH}/test_weather.csv")
+    commit = tx.commit(
+        repository=REPO_NAME,
+        branch=TRAINING_BRANCH,
+        message="Perform train-test split of 2010 weather data",
+    )
+    tx.tag(repository=REPO_NAME, ref=commit, tag=TAG_NAME)
+
+# %% [markdown]
+"""
+Now we have the data on different branches. If new data comes in, we can perform necessary preprocessing on a separate branch and merge it to `main` once we are sure about its compatibility and we have run all the necessary tests.
+Should the new data be important for the experimentation as well, then we can merge the new main branch into the experimentation branch.
+You should then create a new tag for the dataset.
+You cannot directly reassign tags. If you want to do this anyways, for example to prevent namespace clutter, you have to delete the tag and create a new one.
+However, beware as this might break reproducibility in other places (i.e. colleagues might expect unchanged data).
+To ensure failsafe versioning use the SHA's of the commits in tracking tools.
```

Comment:
This is not entirely clear to me - you are saying that branch names as refs are shallow, right? Meaning, they can point to different versions depending on the repo state.

Good point, but that should be communicated more efficiently, something as in "To ensure full reproducibility, only use actual commit SHAs instead of branch or tag names in experiment tracking tools."

Reply:
Jap. Thats it. Reworded.